crates/core/component/stake/src/component/validator_handler/uptime_tracker.rs

```diff
@@ -0,0 +1,189 @@
+use {
+    super::{ValidatorDataRead, ValidatorDataWrite, ValidatorManager},
+    crate::{
+        component::{
+            metrics,
+            stake::{
+                address::{validator_address, Address},
+                ConsensusIndexRead,
+            },
+            StateReadExt as _,
+        },
+        params::StakeParameters,
+        validator, IdentityKey, Uptime,
+    },
+    anyhow::Result,
+    async_trait::async_trait,
+    cnidarium::StateWrite,
+    futures::StreamExt as _,
+    penumbra_sct::component::clock::EpochRead,
+    std::collections::BTreeMap,
+    tap::Tap,
+    tendermint::abci::types::CommitInfo,
+    tokio::task::{AbortHandle, JoinSet},
+    tracing::{debug, error_span, instrument, trace, Instrument},
+};
+
+/// A bundle of information about a validator used to track its uptime.
+type ValidatorInformation = (IdentityKey, tendermint::PublicKey, Uptime);
+
+/// A collection of tasks retrieving [`ValidatorInformation`].
+type Lookups = JoinSet<anyhow::Result<Option<ValidatorInformation>>>;
+
+/// Tracks validator uptimes.
+///
+/// Use [`track_uptime()`] to process a block's [`CommitInfo`] and update validator uptime
+/// bookkeeping.
+///
+/// [`track_uptime()`]: Self::track_uptime
+#[async_trait]
+pub trait ValidatorUptimeTracker: StateWrite {
+    #[instrument(skip(self, last_commit_info))]
+    async fn track_uptime(&mut self, last_commit_info: &CommitInfo) -> Result<()> {
+        // Note: this probably isn't the correct height for the LastCommitInfo,
+        // which is about the *last* commit, but at least it'll be consistent,
+        // which is all we need to count signatures.
+        let height = self.get_block_height().await?;
+        let params = self.get_stake_params().await?;
+
+        // Build a mapping from addresses (20-byte truncated SHA256(pubkey)) to vote statuses.
+        let did_address_vote = last_commit_info
+            .votes
+            .as_slice()
+            .tap(|votes| {
+                if votes.is_empty() {
+                    debug!("no validators voted")
+                } else {
+                    debug!(len = %votes.len(), "collecting validator votes")
+                }
+            })
+            .into_iter()
+            .map(|vote| (vote.validator.address, vote.sig_info.is_signed()))
+            .inspect(|(address, voted)| {
+                trace!(
+                    address = %hex::encode(address),
+                    %voted,
+                    "validator vote information"
+                )
+            })
+            .collect::<BTreeMap<Address, bool>>();
+
+        // Since we don't have a lookup from "addresses" to identity keys,
+        // iterate over our app's validators, and match them up with the vote data.
+        // We can fetch all the data required for processing each validator concurrently:
+        let mut lookups = Lookups::new();
```

Comment:
nit: imo, the aliasing/varname makes the code a little bit harder to read, this is a `JoinSet` but I'd have to keep on reading to the end of method to learn that.

Reply:
for sure. i'd hoped putting `Lookups` at the top of the file was declaring it up front, but i think that you are right.

how does addb311e21f3420d5a3b5ca3f85c9d9499fc82fe (see #4102) look to you? that promotes `JoinSet` out of the type alias, so we know what `lookups` _is_, and uses the alias to instead refer to the output of those tasks. cake, and eating it too. :cake: :fork_and_knife: :yum: